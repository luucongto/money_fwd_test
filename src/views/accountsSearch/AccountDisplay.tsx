import {
  CButton,
  CCard,
  CCardBody, CCardHeader, CDataTable, CModal, CModalBody, CModalFooter, CModalHeader, CModalTitle
} from '@coreui/react'
import { useEffect, useState } from 'react'
import { useAppDispatch, useAppSelector } from '../../redux/hooks'
import { actions } from '../../redux/reducers/user'

const AccountDisplay = (props: any): JSX.Element => {
  const dispatch = useAppDispatch()
  const userId = props?.match?.params?.userId
  const [modal, setModal] = useState(false)
  useEffect(() => {
    dispatch(actions.userRequest({ userId }))
  }, [userId, dispatch])
  const userData = useAppSelector(state => { return state.user.data != null ? state.user.data : { userId, accounts: [] } })
  const isLoading = useAppSelector(state => state.user.fetching)
  const error = useAppSelector(state => { return state.user.error != null ? state.user.error : null })
  useEffect(() => {
    if (error !== null) {
      setModal(true)
    }
  }, [error])
  const fields = ['id', 'name', 'balance']
  return (
    <>
      <CCard>
        <CCardHeader>
          UserId: {userData?.userId}
        </CCardHeader>
        <CCardBody>
          {isLoading ? 'Loading'
            : (
              <CDataTable
                items={userData?.accounts}
                fields={fields}
                itemsPerPage={10}
                pagination
              />
            )}

        </CCardBody>
      </CCard>

      <CModal
        show={modal}
        onClose={setModal}
      >
        <CModalHeader closeButton>
          <CModalTitle>Error</CModalTitle>
        </CModalHeader>
        <CModalBody>
          {error?.toUpperCase()}
        </CModalBody>
        <CModalFooter>
          <CButton
            color='primary'
            onClick={() => {
              setModal(false)
              dispatch(actions.clearError())
            }}
          >Close
          </CButton>
        </CModalFooter>
      </CModal>
    </>
  )
}

export default AccountDisplay
