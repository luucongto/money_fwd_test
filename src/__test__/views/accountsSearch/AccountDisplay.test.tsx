import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { useAlert } from 'react-alert'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import createSagaMiddleware from 'redux-saga'
import AccountDisplay from '../../../views/accountsSearch/AccountDisplay'

const sagaMiddleware = createSagaMiddleware()
const mockStore = configureStore([sagaMiddleware])

jest.mock('react-alert', () => ({
  ...jest.requireActual('react-alert'),
  useAlert: () => ({
    show: jest.fn()
  })
}))

describe('AccountDisplay display user accounts', () => {
  let wrapperResult: ReactWrapper
  let store: any
  beforeEach(() => {
    const initState = {
      user: {
        error: null,
        fetching: false,
        data: null
      }
    }

    store = mockStore(initState)
    wrapperResult = mount(
      <Provider store={store}>
        <AccountDisplay match={{ params: { userId: 1 } }} />
      </Provider>
    )
  })

  test('renders AccountDisplay without crashing', () => {
    expect(wrapperResult).toMatchSnapshot()
    expect(wrapperResult.text().includes('UserId: 1')).toBe(true)
  })
})

describe('AccountDisplay display user', () => {
  let wrapperResult: ReactWrapper
  let store: any
  beforeEach(() => {
    const initState = {
      user: {
        error: 'User Not Found',
        fetching: false,
        data: null
      }
    }

    store = mockStore(initState)
    wrapperResult = mount(
      <Provider store={store}>
        <AccountDisplay />
      </Provider>
    )
  })

  test('renders AccountDisplay without error', () => {
    expect(store.getActions().length).toEqual(2)
    expect(store.getActions()[1].type).toEqual('user/clearError')
  })
})
