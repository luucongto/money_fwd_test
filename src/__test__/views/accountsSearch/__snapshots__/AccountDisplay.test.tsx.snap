// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDisplay display user accounts renders AccountDisplay without crashing 1`] = `ReactWrapper {}`;

exports[`AccountDisplay display user renders AccountDisplay without crashing 1`] = `ReactWrapper {}`;
